streaming/src/main/scala/org/apache/gearpump/streaming/appmaster/TaskLocator.scala

```diff
@@ -20,70 +20,55 @@ package org.apache.gearpump.streaming.appmaster
 import java.util
 
 import akka.actor.Actor
-import com.typesafe.config.Config
+import com.typesafe.config.{ConfigRenderOptions, Config}
 import org.apache.gearpump.cluster.ClusterConfig
-import org.apache.gearpump.streaming.ProcessorDescription
-import org.apache.gearpump.streaming.appmaster.TaskLocator.{WorkerLocality, NonLocality, Locality}
-import org.apache.gearpump.streaming.task.{Task, TaskUtil}
+import org.apache.gearpump.streaming.{ProcessorId, ProcessorDescription}
+import org.apache.gearpump.streaming.appmaster.TaskLocator.{Localities, WorkerLocality, NonLocality, Locality}
+import org.apache.gearpump.streaming.task.{TaskId, Task, TaskUtil}
 import org.apache.gearpump.util.{ActorUtil, Constants}
 
 import scala.collection.mutable
+import scala.util.Try
 
-class TaskLocator(config: Config) {
-  private var userScheduledTask = Map.empty[Class[_ <: Task], mutable.Queue[Locality]]
+class TaskLocator(appName: String, config: Config) {
+  private var taskLocalities: Map[TaskId, Locality] = loadTaskLocalities(config)
 
-  initTasks()
-
-  def initTasks() : Unit = {
-    val taskLocations : Array[(ProcessorDescription, Locality)] = loadUserAllocation(config)
-    for(taskLocation <- taskLocations){
-      val (taskDescription, locality) = taskLocation
-      val localityQueue = userScheduledTask.getOrElse(TaskUtil.loadClass(taskDescription.taskClass), mutable.Queue.empty[Locality])
-      0.until(taskDescription.parallelism).foreach(_ => localityQueue.enqueue(locality))
-      userScheduledTask += (TaskUtil.loadClass(taskDescription.taskClass) -> localityQueue)
-    }
-  }
-
-  def locateTask(taskDescription : ProcessorDescription) : Locality = {
-    if(userScheduledTask.contains(TaskUtil.loadClass(taskDescription.taskClass))){
-      val localityQueue = userScheduledTask.get(TaskUtil.loadClass(taskDescription.taskClass)).get
-      if(localityQueue.size > 0){
-        return localityQueue.dequeue()
-      }
-    }
-    NonLocality
+  def locateTask(taskId: TaskId) : Locality = {
+    taskLocalities.getOrElse(taskId, NonLocality)
   }
 
   /*
 The task resource requests format:
 gearpump {
-  scheduling {
-    requests {
-      worker1 {
-        "task1" = 2   //parallelism
-        "task2" = 4
+  streaming {
+    localities {
+      app1: {
+        workerId: [
+          TaskId(0,0), TaskId(0,1)
+        ]
       }
-      worker2 {
-        "task1" = 2
+      app2: {
+        workerId: [
+          TaskId(1,0), TaskId(1,1)
+        ]
       }
```

Comment:
1. Mostly, the data is stored by AppMaster automatically, so end user don't need to understand much details, so that it can keep its locality when there is AppMaster recovery happen.

2. If user want to change it theirselves, this can be easily done, as user can do a easy query to get the id and the parallism of each processor.